### PR TITLE
cli: add anchor clean

### LINKED ISSUE #1470: cli: add anchor clean
Add an `anchor clean` command equivalent to `cargo clean` but preserving `./target/deploy/program-key.json` files

### PR DESCRIPTION
**Pull request by skrrb**
Closes #1470

### CODE REVIEW

File: cli/src/lib.rs [hunk 1]
```diff
@@ -2100,6 +2103,31 @@ fn cluster_url(cfg: &Config) -> String {
...
+        if path.is_dir() && path != deploy_dir {
+            fs::remove_dir_all(path).expect("Could not remove directory.");
+        } else if path.is_file() {
+            fs::remove_file(path).expect("Could not remove file.");
+        }
+    }
+
+    for file in fs::read_dir(deploy_dir)? {
+        let path = file?.path();
+        if !path.to_string_lossy().ends_with("-keypair.json") {
```
Comment (paul-schaaf) [inline cli/src/lib.rs:2123]: it's probably unlikely people will rename their json files here but to be safe, what do you think about changing this to just ".json"

File: cli/src/lib.rs [hunk 2]
```diff
@@ -2100,6 +2103,31 @@ fn cluster_url(cfg: &Config) -> String {
...
+fn clean(cfg_override: &ConfigOverride) -> Result<()> {
+    let cfg = Config::discover(cfg_override)?.expect("Not in workspace.");
+    let cfg_parent = cfg.path().parent().expect("Invalid Anchor.toml");
+    let target_dir = cfg_parent.join("target");
+    let deploy_dir = target_dir.join("deploy");
+
+    for entry in fs::read_dir(target_dir)? {
+        let path = entry?.path();
+        if path.is_dir() && path != deploy_dir {
+            fs::remove_dir_all(path).expect("Could not remove directory.");
```
Comment (paul-schaaf) [inline cli/src/lib.rs:2115]: pls provide as much helpful info as possible in error msgs. In this case, include the path of the dir that could not be removed

File: cli/src/lib.rs [hunk 3]
```diff
@@ -181,6 +181,8 @@ pub enum Command {
         #[clap(subcommand)]
         subcmd: IdlCommand,
     },
+    /// Remove generated artifacts.
```
Comment (paul-schaaf) [inline cli/src/lib.rs:184]: also say what is *not* being deleted

File: cli/src/lib.rs (hunk 1 above)
Comment (skrrb) [reply cli/src/lib.rs:2123]: yeah, better to consider those cases as well. Maybe it's worth also checking that the .json is a valid keypair?

File: cli/src/lib.rs (hunk 3 above)
Comment (skrrb) [reply cli/src/lib.rs]: Changed to `Remove all artifacts from the target directory except program keypairs`

File: cli/src/lib.rs (hunk 2 above)
Comment (skrrb) [reply cli/src/lib.rs]: added a warning with the failed file/directory to be deleted and the error. I think it is better not to panic as the following files could be deleted successfully

Comment (paul-schaaf) [reply cli/src/lib.rs]: I actually think panicing is the better thing to do here.

if an error happens here something unusual is happening that should be investigated and also should make CI flows fail.

File: cli/src/lib.rs (hunk 1 above)
Comment (paul-schaaf) [reply cli/src/lib.rs]: dont think thats necessary. If someone put a json file there that is _not_ a valid keypair then thats probably intentional and we shouldnt delete that either

File: cli/src/lib.rs (hunk 2 above)
Comment (skrrb) [reply cli/src/lib.rs]: yes, I hadn't thought of it that way. I have added back the error (but more verbose)

**Review by paul-schaaf — APPROVED** [review]

**Comment by paul-schaaf** [comment]
could u sign your previous commits and force push?

**Comment by paul-schaaf** [comment]
@skrrb u need to sign your previous commits too

**Comment by skrrb** [comment]
> @skrrb u need to sign your previous commits too

yes, I'm trying to do it but tbh I really don't know how to do it. 

I tried to sign them with `git filter-branch`, but although it works for my commits, for the rest the signature is lost. with an interactive rebase I can't do it either

**Comment by paul-schaaf** [comment]
we've turned off the signature requirement. Thanks for the contribution!

**Pull request merged**